### ISSUE TITLE
undo/redo support

### ISSUE THREAD

**Issue by batwicket** [matthiasn/systems-toolbox#14]
I'm getting to the point where undo/redo is handy. Tixi (https://github.com/astashov/tixi/blob/c1fae3c7ff364895fe3371eba336ddf8f6a9d6e0/README.md) uses a tree/zipper, which allows the sequence:
<undo> <change> <redo>
The standard flat undo/redo is ok for my needs, though.

That said, many of the React framework examples keep state centrally, which makes this feature trivial. I'm not sure what your plans are for this so i haven't done a pull request.

**Comment by matthiasn** [comment]
In general, this would be trivial achieve, I just don't have a great where to store component state yet. On the client side, it may seem like a good idea to have a central store, but I'd rather keep state management local to the component as some centralized data store may become a source of contention when building a system in a multi-threaded environment. This issue should be thought through together with issue #15.

**Comment by batwicket** [comment]
Yes, multi-threaded mobile devices could become a viable target for the toolbox.

Perhaps the approach taken at https://github.com/reagent-project/historian/blob/master/src/cljc/historian/core.cljc is of interest?

**Comment by matthiasn** [comment]
Yes that would be interesting. Looks like historian doesn't tie us to reagent?

I was mostly thinking about the server side on the JVM here. But the messaging model may also enable a threaded environment in the browser by leveraging web workers. That could be really interesting, actually.

**Comment by batwicket** [comment]
I think Historian's approach is not reagent-specific. It seems to be consistent with the way Clojurescript handles state.

Perhaps server-side undo/redo isn't so important these days. The ability to run the GUI offline is often nice and many server-side services don't easily support interactive undo/redo. Just saying.. maybe you have a case where server-side undo is important?

**Comment by matthiasn** [comment]
Oh, what I wrote was probably a bit ambiguous. I was talking about the multi-threaded environment when mentioning the JVM.
The undo/redo that you mentioned is something that I would only see on the client indeed. Undo on the server would look very different, likely involve rolling back database transactions and stuff. That would almost certainly be so application-specific that I doubt there could be a way to extract a useful generalized approach.
